exasol_transformers_extension/udfs/models/token_classification_udf.py

```diff
@@ -0,0 +1,199 @@
+import torch
+import pandas as pd
+import transformers
+from typing import Tuple, List, Union, Dict
+from exasol_transformers_extension.deployment import constants
+from exasol_transformers_extension.utils import device_management, \
+    dataframe_operations, bucketfs_operations
+
+
+class TokenClassificationUDF:
+    def __init__(self,
+                 exa,
+                 batch_size=100,
+                 pipeline=transformers.pipeline,
+                 base_model=transformers.AutoModelForTokenClassification,
+                 tokenizer=transformers.AutoTokenizer):
+        self.exa = exa
+        self.bacth_size = batch_size
+        self.pipeline = pipeline
+        self.base_model = base_model
+        self.tokenizer = tokenizer
+        self.device = None
+        self.cache_dir = None
+        self.last_loaded_model_key = None
+        self.last_loaded_model = None
+        self.last_loaded_tokenizer = None
+        self.last_created_pipeline = None
+        self.last_used_top_k = None
+        self.default_aggregation_strategy = 'simple'
+
+    def run(self, ctx):
+        device_id = ctx.get_dataframe(1).iloc[0]['device_id']
+        self.device = device_management.get_torch_device(device_id)
+        ctx.reset()
+
+        while True:
+            batch_df = ctx.get_dataframe(num_rows=self.bacth_size, start_col=1)
+            if batch_df is None:
+                break
+
+            result_df = self.get_batched_predictions(batch_df)
+            ctx.emit(result_df)
+
+        self.clear_device_memory()
+
+    def get_batched_predictions(self, batch_df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Perform separate predictions for each model in the dataframe. If the
+        model is not cached, it is loaded into the cache before the prediction.
+
+        :param batch_df: A batch of dataframe retrieved from context
+
+        :return: Prediction results of the corresponding dataframe
+        """
+        result_df_list = []
+        unique_values = dataframe_operations.get_unique_values(
+            batch_df, constants.ORDERED_COLUMNS, sort=True)
+        for model_name, bucketfs_conn, sub_dir in unique_values:
+            model_df = batch_df[
+                (batch_df['model_name'] == model_name) &
+                (batch_df['bucketfs_conn'] == bucketfs_conn) &
+                (batch_df['sub_dir'] == sub_dir)]
+
+            current_model_key = (bucketfs_conn, sub_dir, model_name)
+            if self.last_loaded_model_key != current_model_key:
+                self.set_cache_dir(model_df)
+                self.clear_device_memory()
+                self.load_models(model_name)
+                self.last_loaded_model_key = current_model_key
+                self.last_used_top_k = None
```

Comment:
I think, we don't need top_k for token classification

Reply:
This variable is leftover. We dont have topk for this task